### ISSUE TITLE
[EPIC] Get data for calculated fields

### ISSUE THREAD

**Issue by tmetzke** [camunda/camunda#22812]
## Description

For some entities that we want to expose in the C8 REST API, some fields are not directly accessible from one entity in one index.
This epic collects the entities and fields in question to decide how we deal with them.

## Options

* Not expose the fields in the API
* Fetch from multiple indexes in the search layer implementation
* Provide the data in the indices as needed directly

## Entities and fields

- [ ] https://github.com/camunda/camunda/issues/21945

**Comment by tmetzke** [comment]
Public processes:
* Secondary priority for 8.7, should-have rather than a must-have
* We need to add `isPublic` to the `*-process` index

Process definition form:
* We need to add `formKey` to the `*-process` index

Flow node name:
* We'll check usage, e.g. in Operate frontend (@ralfpuchert).
* If it's **just** in the API but not used in our frontend, we'll not deliver this field anymore in V2.
* We'll check in C7 API as well (@tmetzke)
* Alternative/workaround for V2: you have to fetch the XML separately in a second call from the client

Incident <-> flow node instance:
* There should only be one incident per flow node instance
* Execution and task listeners adhere to this as well (handled sequentially)
* We'll check with the Zeebe teams regarding parallel multi-instance execution (@tmetzke)

**Comment by ralfpuchert** [comment]
@tmetzke flow node name is only used in v1. So we don't deliver this in V2.

**Comment by tmetzke** [comment]
> @tmetzke flow node name is only used in v1. So we don't deliver this in V2.

So, in Operate, we fetch the XML separately currently and resolve the names of the flow nodes in the client application?

**Comment by tmetzke** [comment]
> @tmetzke flow node name is only used in v1. So we don't deliver this in V2.

Thanks, @ralfpuchert. So, in Operate, we fetch the XML separately currently and resolve the names of the flow nodes in the client application?

**Comment by tmetzke** [comment]
Regarding incidents, I am confident that each incident will be attached to the inner multi-instance activity instances, not the outer body. It is a 1:1 relation even for those scenarios.
I debugged the [MultiInstanceIncidentTest](https://github.com/camunda/camunda/blob/052efc8d7c27b0be3c956c9580f66b58acfe6e96/zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/incident/MultiInstanceIncidentTest.java#L270-L326), all incidents are attached to different element instances.
![image](https://github.com/user-attachments/assets/27bd381b-ad4b-4299-ba53-69c3d46906a9)

**Comment by houssain-barouni** [comment]
@tmetzke FYI, I added 
[Decision definition](https://github.com/camunda/camunda/issues/20654): `decisionRequirementsName`, `decisionRequirementsVersion` - `operate-decision` index

**Comment by ralfpuchert** [comment]
> > @tmetzke flow node name is only used in v1. So we don't deliver this in V2.
> 
> Thanks, @ralfpuchert. So, in Operate, we fetch the XML separately currently and resolve the names of the flow nodes in the client application?

Yes, frontend parses the XML and retrieves the flow node names from there. Ironically backend does something similar in order to provide it for v1.

**Comment by tmetzke** [comment]
> > > @tmetzke flow node name is only used in v1. So we don't deliver this in V2.
> > 
> > 
> > Thanks, @ralfpuchert. So, in Operate, we fetch the XML separately currently and resolve the names of the flow nodes in the client application?
> 
> Yes, frontend parses the XML and retrieves the flow node names from there. Ironically backend does something similar in order to provide it for v1.

Thanks, @ralfpuchert, that helps a lot to understand this. OK, we'll get rid of the backend cache for the C8 REST API at least then 🙂

**Comment by tmetzke** [comment]
@ralfpuchert, added you as the DRI here to overlook that all required changes are handled by devs accordingly.

**Comment by tmetzke** [comment]
> https://github.com/camunda/camunda/issues/20652: isPublic - operate-process index (Check with web modeler team)

Will be removed from the API

**Comment by tmetzke** [comment]
Regarding attributes like `flowNodeName` and `taskName`, resolved from BPMN XML data, @Zelldon and @tmetzke will sync on a solution proposal.

Requirement:
* The data should be available via API 

Reasoning:
* The client applications must otherwise fetch the XML to resolve those fields.
* However, the user interaction and the authenticated token associated with it might not have the required permissions to fetch those XMLs.

Options:
* The client applications own an "internal" token/account as well that has the required permissions to fetch XMLs.
  * Con: this needs to be maintained as an account in the cluster, otherwise it's some "special treatment" request which makes the permission handling too complex and error-prone.
* The REST API resolves the fields by fetching from multiple indices and caching the BPMN (XMLs or meta-models) to avoid fetching too often.
* The Camunda exporter writing the aggregated data writes the attributes.

**Comment by Zelldon** [comment]
Let us sync tomorrow

**Comment by tmetzke** [comment]
## Decision on BPMN XML-related attributes


### What is the proposal?

-   The REST API will take care of resolving such attributes by itself. This will not be provided by the data layer (ES, OS, RDBMS).
-   The REST API will cache BPMN XMLs (or processed versions of them) to resolve this data to the required attributes.
-   The REST API will fetch missing BPMN XMLs into that cache by requesting the data from the data layer.

### Why do we do it that way?

-   The client applications cannot conveniently resolve this data themselves
    -   They would need to ensure they can fetch the data with an appropriate token/account that has the right permissions in Identity.
    -   This would need to exist separately to the user token that requests the user-facing entity, e.g. a list of user tasks.
    -   That user token itself might not have the right permissions to fetch the XMLs.
-   The data layer seems like a good fit to provide this data but it might not really be
    -   The ES/OS exporter would need some substantial reworking to allow for a consistent process cache. That work is currently not planned for 8.7 and would put those attribute completely at risk for 8.7.
    -   The RDBMS exporter would also need to implement this and might have to adjust this for every supported database.
    -   If we want to support further data layers, those would need to implement this as well.
    -   When a new attribute that is resolved from BPMN XML data is introduced for an entity, say, in 8.8, we have to decide whether this data is also available for older data entries
        -   If not, the UX for users might be a bit weird since they can only see data for that attribute in the API for newly created instances but not for older ones although the data also exists for those in the BPMN.
        -   If so, we would have to migrate existing data to also contain this attribute by resolving this from the BPMN XML in a data migration routine.
-   In the REST API, there are a couple of benefits
    -   We can conveniently fetch the XML data without authorization context to ensure we can receive it.
    -   We can build it once and reuse it almost regardless of the data layer underneath (we have to implement a query for fetching XMLs anyway for the related API).
    -   We can easily resolve new attributes for existing data instances by adding a routine to fill a new attribute with XML-based data at runtime. No need for data migration.

### What are the downsides?

-   REST API queries might be slower depending on the existing cache entries. Missing XMLs have to be fetched first.
-   An unexpected work item for the REST API team is added. Due to dropping historic resource deletion (see post above), we have some capacity though to tackle this instead directly.

### Next steps

- @tmetzke creates an implementation epic for the REST API-resolution of BPMN XML-based attributes
- @ralfpuchert and @tmetzke sync on implementation effort and timebox for that epic

**Comment by tmetzke** [comment]
@houssain-barouni and @ralfpuchert, regarding
> - [ ] [Decision definition](https://github.com/camunda/camunda/issues/20654): `decisionRequirementsName`, `decisionRequirementsVersion` - `operate-decision` index

Where did we get this data from previously for V1 API?

**Comment by houssain-barouni** [comment]
> @houssain-barouni and @ralfpuchert, regarding
> 
> > * [ ]  [Decision definition](https://github.com/camunda/camunda/issues/20654): `decisionRequirementsName`, `decisionRequirementsVersion` - `operate-decision` index
> 
> Where did we get this data from previously for V1 API?

@tmetzke in v1, this data was retrieved from the `operate-decision-requirements` index, requiring an additional call to Elasticsearch to fetch the decision requirements for the decision definitions matching the search request, to populate these fields in the decision definition search response

**Comment by tmetzke** [comment]
This one has been flying under my radar, I just noticed.
@marcosgvieira, we need to investigate the missing attributes in the [decision definition search endpoint](https://docs.camunda.io/docs/next/apis-tools/camunda-api-rest/specifications/search-decision-definitions/) compared to the [V1 endpoint](https://docs.camunda.io/docs/next/apis-tools/operate-api/specifications/search-7/):
 
* We're missing `decisionRequirementsName` and `decisionRequirementsVersion` as search filter criteria.
* They are not stored in the index we use for the decision definition search itself. Searching in two indices can break pagination and such.

We need to clarify on a technical level if we can achieve the same coverage and what we do in Operate V1. Depending on that, we need to decide if we want to add those attributes as filter criteria in V2 as well or if they are a known and consciously accepted regression from (Operate) V1 to V2 API.

**Comment by marcosgvieira** [comment]
There is a related issue with some analysis already: https://github.com/camunda/camunda/issues/20767

**Comment by kristinkomschow** [comment]
Unsure who to tag, but can we close his issue since we already have another one for the remaining Decision Definition which seems like it was the one open task left?

@marcosgvieira @eppdot I heard you are planning on screening remaining regressions and collect them under one epic, so it seems like this issue will be redundant.

**Comment by tmetzke** [comment]
I'm all for closing this based on this 👍 

> Unsure who to tag, but can we close his issue since we already have another one for the remaining Decision Definition which seems like it was the one open task left?
> 
> [@marcosgvieira](https://github.com/marcosgvieira) [@eppdot](https://github.com/eppdot) I heard you are planning on screening remaining regressions and collect them under one epic, so it seems like this issue will be redundant.

**Comment by eppdot** [comment]
We have a separate follow up issue for the decision definition search #20767 
Thus closing this one.
